pkg/reconciler/pipelinerun/resources/validate_params.go

```diff
@@ -110,3 +113,42 @@ func ValidateParameterTypesInMatrix(state PipelineRunState) error {
 	}
 	return nil
 }
+
+// ValidateParamArrayIndex validates if the param reference to an array param is out of bound.
+// error is returned when the array indexing reference is out of bound of the array param
+// e.g. if a param reference of $(params.array-param[2]) and the array param is of length 2.
+func ValidateParamArrayIndex(ctx context.Context, ps *v1beta1.PipelineSpec, pipelineRunParams v1beta1.Params) error {
+	// Collect all array params lengths
+	arrayParamsLengths := ps.Params.ExtractDefaultParamArrayLengths()
+	for k, v := range pipelineRunParams.ExtractParamArrayLengths() {
+		arrayParamsLengths[k] = v
+	}
+	// extract all array indexing references, for example []{"$(params.array-params[1])"}
+	arrayIndexParamRefs := ps.GetIndexingReferencesToArrayParams().List()
+
+	return validateOutofBoundArrayParams(arrayIndexParamRefs, arrayParamsLengths)
+}
+
+// validateOutofBoundArrayParams validates if the array indexing params are out of bound
+// example of arrayIndexingParams: ["$(params.a-array-param[1])", "$(params.b-array-param[2])"]
+// example of arrayParamsLengths: {"a-array-params": 2,"b-array-params": 2 }
+func validateOutofBoundArrayParams(arrayIndexingParams []string, arrayParamsLengths map[string]int) error {
+	outofBoundParams := sets.String{}
+	for _, val := range arrayIndexingParams {
+		indexString := substitution.ExtractIndexString(val)
+		idx, _ := substitution.ExtractIndex(indexString)
+		// this will extract the param name from reference
+		// e.g. $(params.a-array-param[1]) -> a-array-param
+		paramName, _, _ := substitution.ExtractVariablesFromString(substitution.TrimArrayIndex(val), "params")
+
+		if paramLength, ok := arrayParamsLengths[paramName[0]]; ok {
+			if idx >= paramLength {
+				outofBoundParams.Insert(val)
+			}
+		}
+	}
+	if outofBoundParams.Len() > 0 {
+		return fmt.Errorf("non-existent param references:%v", outofBoundParams.List())
```

Comment:
Sorry I know this may not be a change we need to have in this PR, but should we have more specific error message here? I don't remember why I use "non-existent param references", does it make more sense to say that those params are out of bound? If so I can fix this in another PR.

Reply:
If you wouldn't mind fixing this in a separate PR that would be great!